### ISSUE TITLE
Old versions yanked (e.g. 0.3.5)

### ISSUE THREAD

**Issue by makyen** [minad/mimemagic#148]
It appears all older versions of this gem have been yanked from distribution. IMO, that's generally inappropriate. While it's beneficial for people to update to new versions, that's something which each project which uses the gem should be permitted to determine when, or even if, they want to do so. Generally, yanking old versions should only be done when the old version is *critically* or *fatally* flawed, which is very rarely the case for packages which were the most current released version for an extended period of time. [Yes, the version may have some significant issues, but those are issues people have been living with for quite some time.]

It might be more reasonable to yank old versions if the new version is actually a direct drop-in replacement for the old version, but that's not the case for this project, because your dependencies updated between those old versions and any version which is currently available.

There are 28,142,644 downloads of 0.3.5. While each one of those downloads isn't going to be a separate project, a large number of them will be. Yanking the old versions basically *forces* every project which is using your gem to do work **now** in order build their project, even if nothing substantial changed in their code.

The decision of when to update dependencies, such as this package, should be left up to the project using it. It shouldn't be dictated by the dependency, except in cases where a critical flaw is found in a recently released version (e.g. someone releases v5.7.3 on day 1, discovers a fatal flaw and releases v5.7.4 on day 2 or 3; then it's reasonable to yank v5.7.3), or when private information is inadvertently published, etc.

**Comment by makyen** [comment]
Looking at the [closed issues](https://github.com/mimemagicrb/mimemagic/issues?q=is%3Aissue+is%3Aclosed) on this project, I see that there are at least a couple/few closed issues about the older versions being yanked.

I see from [this comment](https://github.com/mimemagicrb/mimemagic/issues/127#issuecomment-807902889) by [rstacruz](https://github.com/rstacruz) that the issue is a licensing problem:

> Unfortunately this [keeping old versions] isn't possible because of legal licensing issues. It also isn't possible to push new code in place of the old versions.
> 
> * https://www.theregister.com/2021/03/25/ruby_rails_code/
> 
> * [rails/rails#41750](https://github.com/rails/rails/issues/41750)

**Comment by makyen** [comment]
Given that there are going to be a substantial number of people investigating why their builds are breaking, and the error will point them at this project, I suggest leaving open at least one of the Issues which covers the problem and at least provides links to further information. It doesn't need to be this specific issue. It doesn't need to be any of the ones which have already been created.

However, it would be good to have one open Issue, so that people will more easily find it and be able to:

* find an explanation as to why the old versions were yanked
* see that yanking the older versions was *necessary*
* have information or links to information as to how things can be immediately resolved for a project which was using this gem
* find out about the potentially significant issue that their project may have included GPL code.
  * That their project contained GPL code, even inadvertently, imposes legal requirements upon them. They need to be informed that the issue exists, so they can deal with it in whatever way they find necessary. For some projects, this may require getting their lawyers involved in determining how they want to deal with the issue.

**Comment by jm4rsh4ll** [comment]
Quite sad to see it. I stuck because of this change for entire day with my, quite big and old project which I maintain alone. Why the next version of this gem depends on the latest nokogiri? I don't know how to resolve all conflicts now.

🙁

**Comment by denodster** [comment]
You all really need to figure out a way for this to stop, At first I thought it was going to be a one time thing, but now I see that you think you're special and that we're willing to put up with this. I'm going to start looking for a way to dump this gem if this continues. At some point I'm going to push code to production and find out that deployment fails because mimemagic was updated at some point between when we started our last patch and when we shipped.
